src/Vulkan/Silk.NET.Vulkan/Vk.cs

```diff
@@ -197,18 +225,69 @@ public bool IsDeviceExtensionPresent(Instance instance, string extension)
+            // as the operations are inheritently safe, the additional try...finally on write is mitigated
+            // as we also use it to dispose the memory allocation safely.  This gives us thread-safety at
+            // no real additional cost.
+            _cachedDeviceExtensionsLock.EnterUpgradeableReadLock();
+
+            // We check for the extension first to avoid 2 lookups
+            if (_cachedDeviceExtensions.Contains(fullKey))
             {
-                var extProperties = stackalloc ExtensionProperties[128];
-                _cachedDeviceExtensions.Add(device.Handle, list = Add(device, new List<string>(), extProperties));
+                // We found the extension
+                result = true;
             }
-            else
+            else if (!_cachedDeviceExtensions.Contains(prefix))
             {
-                list = _cachedDeviceExtensions[device.Handle];
-            }
+                // The lack of the device handle indicates we've not been previously initialised.  We now need a write lock.
+                _cachedDeviceExtensionsLock.EnterWriteLock();
+                GlobalMemory mem = null;
+                try
+                {
+                    var deviceExtPropertiesCount = 0u;
+
+                    // Get number of properties
+                    EnumerateDeviceExtensionProperties(device, (byte*) 0, &deviceExtPropertiesCount, null);
+
+                    // Initialise return structure
+                    mem = GlobalMemory.Allocate((int) deviceExtPropertiesCount * sizeof(ExtensionProperties));
+                    var props = (ExtensionProperties*) Unsafe.AsPointer(ref mem.GetPinnableReference());
 
-            return list.Contains(extension);
+                    // Get properties
+                    EnumerateDeviceExtensionProperties(device, (byte*) 0, &deviceExtPropertiesCount, props);
+                    for (int j = 0; j < deviceExtPropertiesCount; j++)
+                    {
+                        // Prefix the extension name
+                        var newKey = prefix_sep + Marshal.PtrToStringAnsi((IntPtr) props[j].ExtensionName);
+                        _cachedDeviceExtensions.Add(newKey);
+                        if (!result && string.Equals(newKey, fullKey))
+                        {
+                            // We found the extension (no need to do another lookup as we're scanning anyway)
+                            // As such this has taken 2 lookups + initialisation scan.
+                            result = true;
+                        }
+                    }
+                }
+                finally
+                {
+                    _cachedDeviceExtensionsLock.ExitWriteLock();
+                    mem?.Dispose();
+                }
+            } // else result = false - takes 2 lookups, one to check for extension, and one to check initialisation.
```

Comment:
I'd expect it to be faster to use a separate bool or nullability to indicate initialization. Why the extra lookup?

Reply:
You have to indicate initialisation per device, which is itself unbounded, requiring a hashset.  For an optimisation we use the same hashset as the composite one.  There is an argument that you could hash the IntPtr's separately to indicate initialisation, and, in fact that might be very slightly quicker, but as we need to create the string version for the composite key anyway and the hash lookup of instance occurs only on cache miss, the additional code bloat and complexity (locking) would not make it worth it in practice.

There is always a worst case of 2 lookups - one to see if the device is initialised and one to see if the extension was found.  The naive original code would first lookup to see if initialised (is the key in the Dictionary O(1) this is the same speed as the HashSet lookup), then would initialise (an O(n) operation, scanning) if necessary and then perform a lookup in the list which is O(n).  This code does the hit test first - O(1) - checks initialisation - O(1) - and, if necessary, initialises O(n), but uses the initialization scan to also locate.  As such, this is O(1), except during initialisation.

In real world use, I'd expect the performance of this code to be significantly faster for all but the smallest collection sizes.  Certainly for collections of 100+ items the `HashSet.Contains` is significantly faster than the `List`.  Most importantly the time remains effectively constant, independent of the size of the extensions (save for the initialisation).

Reply:
This shouldn't be a hot path so not too concerned.